### PR TITLE
bugfix for train-test ratio in statistical testing

### PR DESCRIPTION
**Pull request by lambday**
Fixes the bug found by Dmitry as per the SO thread https://stackoverflow.com/questions/48467554/shogun-quadratic-mmd-error-caused-by-varying-train-test-ratio.

### CODE REVIEW

**Comment by lambday** [comment]
@karlnapf Please have a look. It was caused by the lazy-setters of the train-mode and train-test ratio in `QuadraticTimeMMD::compute_statistic`.

File: src/shogun/statistical_testing/internals/DataManager.cpp [hunk 1]
```diff
@@ -250,7 +247,15 @@ bool DataManager::is_train_test_mode() const
 void DataManager::set_train_mode(bool on)
 {
 	if (train_test_mode)
+	{
 		train_mode=on;
+		REQUIRE(fetchers.size()>0, "Features are not set!");
```
Comment (karlnapf) [inline src/shogun/statistical_testing/internals/DataManager.cpp]: Missing newline

File: src/shogun/statistical_testing/internals/mmd/ComputeMMD.h [hunk 2]
```diff
@@ -87,7 +87,12 @@ struct ComputeMMD
 	{
 		ASSERT(m_n_x>0 && m_n_y>0);
 		const index_t size=m_n_x+m_n_y;
-		ASSERT(kernel_matrix.num_rows==size && kernel_matrix.num_cols==size);
+		REQUIRE(kernel_matrix.num_rows==size,
+				"Number of rows from kernel matrix (%d) did not match the size (%d)\n",
```
Comment (karlnapf) [inline src/shogun/statistical_testing/internals/mmd/ComputeMMD.h]: What size? Should tell the user

**Comment by lambday** [comment]
Thanks, @karlnapf. Added the unit-test that would have failed earlier (unequal train-test split). Let me know if we can merge this once Travis is green.

**Comment by lambday** [comment]
There was also a minor bug which was causing some unit-tests fail. Fixed that as well.

**Comment by vigsterkr** [comment]
@lambday appveyor, i.e. msvc fails with: https://ci.appveyor.com/project/vigsterkr/shogun/build/1818#L16654

```
The following tests FAILED:
	245 - unit-KernelSelectionMaxMMD (Failed)
	247 - unit-KernelSelectionMaxCrossValidation (Failed)
```

more details:
```
Note: Google Test filter = KernelSelectionMaxMMD.*
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from KernelSelectionMaxMMD
[ RUN      ] KernelSelectionMaxMMD.linear_time_single_kernel_streaming
[       OK ] KernelSelectionMaxMMD.linear_time_single_kernel_streaming (8 ms)
[ RUN      ] KernelSelectionMaxMMD.quadratic_time_single_kernel_dense
[       OK ] KernelSelectionMaxMMD.quadratic_time_single_kernel_dense (2 ms)
[ RUN      ] KernelSelectionMaxMMD.quadratic_time_single_kernel_dense_unequal_train_test_ratio
C:\projects\shogun\tests\unit\statistical_testing\KernelSelection_unittest.cc(143): error: The difference between selected_kernel->get_width() and 0.25 is 0.1875, which exceeds 1E-10, where
selected_kernel->get_width() evaluates to 0.0625,
0.25 evaluates to 0.25, and
1E-10 evaluates to 1e-10.
[  FAILED  ] KernelSelectionMaxMMD.quadratic_time_single_kernel_dense_unequal_train_test_ratio (2 ms)
[ RUN      ] KernelSelectionMaxMMD.linear_time_weighted_kernel_streaming
1.000000 
e 1.892112
[       OK ] KernelSelectionMaxMMD.linear_time_weighted_kernel_streaming (8 ms)
[----------] 4 tests from KernelSelectionMaxMMD (20 ms total)
[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (22 ms total)
[  PASSED  ] 3 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] KernelSelectionMaxMMD.quadratic_time_single_kernel_dense_unequal_train_test_ratio
 1 FAILED TEST
        Start 246: unit-KernelSelectionMaxTestPower
246/369 Test #246: unit-KernelSelectionMaxTestPower ................................   Passed    0.03 sec
        Start 247: unit-KernelSelectionMaxCrossValidation
247/369 Test #247: unit-KernelSelectionMaxCrossValidation ..........................***Failed    0.05 sec
Note: Google Test filter = KernelSelectionMaxCrossValidation.*
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from KernelSelectionMaxCrossValidation
[ RUN      ] KernelSelectionMaxCrossValidation.quadratic_time_single_kernel_dense
C:\projects\shogun\tests\unit\statistical_testing\KernelSelection_unittest.cc(327): error: The difference between selected_kernel->get_width() and 0.25 is 0.21875, which exceeds 1E-10, where
selected_kernel->get_width() evaluates to 0.031250000000000014,
0.25 evaluates to 0.25, and
1E-10 evaluates to 1e-10.
[  FAILED  ] KernelSelectionMaxCrossValidation.quadratic_time_single_kernel_dense (3 ms)
[ RUN      ] KernelSelectionMaxCrossValidation.linear_time_single_kernel_dense
[       OK ] KernelSelectionMaxCrossValidation.linear_time_single_kernel_dense (26 ms)
[----------] 2 tests from KernelSelectionMaxCrossValidation (29 ms total)
[----------] Global test environment tear-down
[==========] 2 tests from 1 test case ran. (31 ms total)
[  PASSED  ] 1 test.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] KernelSelectionMaxCrossValidation
```

File: tests/unit/statistical_testing/KernelSelection_unittest.cc [hunk 3]
```diff
@@ -108,6 +108,41 @@ TEST(KernelSelectionMaxMMD, quadratic_time_single_kernel_dense)
...
+	}
+	mmd->set_kernel_selection_strategy(KSM_MAXIMIZE_MMD);
+
+	mmd->set_train_test_mode(true);
+	mmd->set_train_test_ratio(train_test_ratio);
+	mmd->select_kernel();
+	mmd->set_train_test_mode(false);
+
+	auto selected_kernel=static_cast<CGaussianKernel*>(mmd->get_kernel());
+	EXPECT_NEAR(selected_kernel->get_width(), 0.25, 1E-10);
```
Comment (karlnapf) [inline tests/unit/statistical_testing/KernelSelection_unittest.cc]: why is this in here? I think this test is simply about the uneven ratios no?

File: tests/unit/statistical_testing/KernelSelection_unittest.cc [hunk 4]
```diff
@@ -108,6 +108,41 @@ TEST(KernelSelectionMaxMMD, quadratic_time_single_kernel_dense)
...
+	for (auto i=0, sigma=-5; i<num_kernels; ++i, sigma+=1)
+	{
+		float64_t tau=pow(2, sigma);
+		mmd->add_kernel(new CGaussianKernel(10, tau));
+	}
+	mmd->set_kernel_selection_strategy(KSM_MAXIMIZE_MMD);
+
+	mmd->set_train_test_mode(true);
+	mmd->set_train_test_ratio(train_test_ratio);
+	mmd->select_kernel();
```
Comment (karlnapf) [inline tests/unit/statistical_testing/KernelSelection_unittest.cc]: I would just put `EXPECT_NO_THROW` around this and then don't do more.

**Comment by karlnapf** [comment]
The windows failures need fix before this can be merged

**Comment by lambday** [comment]
Thanks, @vigsterkr @karlnapf. Noticed some failures on Travis as well (kernel-selection chooses a different kernel that it was earlier). Need to look into it. Will check and push again.

File: tests/unit/statistical_testing/KernelSelection_unittest.cc (hunk 4 above)
Comment (lambday) [reply tests/unit/statistical_testing/KernelSelection_unittest.cc]: @karlnapf yeah that's better. I'll change it then.

**Comment by lambday** [comment]
Ah sorry, my bad. Both the failures are related to this newly added test. the no-throw thing should fix it.
Also, I am unable to run the unit-test suite locally (with GPL and LIBSVM off). It doesn't compile. Pasting the error next.

**Comment by lambday** [comment]
```
CMakeFiles/shogun-serialization-unit-test.dir/trained_model_serialization_unittest.cc.o:trained_model_serialization_unittest.cc:function TrainedModelSerializationTest_CSVMLin_Test::TestBody(): error: undefined reference to 'shogun::CSVMLin::CSVMLin()
collect2: error: ld returned 1 exit status
make[2]: *** [tests/unit/CMakeFiles/shogun-serialization-unit-test.dir/build.make:294: bin/shogun-serialization-unit-test] Error 1
make[1]: *** [CMakeFiles/Makefile2:393: tests/unit/CMakeFiles/shogun-serialization-unit-test.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 98%] Built target shogun-unit-test
make: *** [Makefile:163: all] Error 2
```

**Comment by karlnapf** [comment]
yeah GPL off stopped working on my machine as well. Just turn it on for now, I will try to fix it later

**Comment by karlnapf** [comment]
I am pretty sure your code will violate the newly enforced style checker.
Run `git clang-format-3.8 --commit 7cfc4b6ffb18574fe5d76fc43c97b997ddbd0163 --binary /usr/bin/clang-format-3.8` with the git revision being the last one in develop before your PR

**Comment by karlnapf** [comment]
Windows is fine now. If you fix the style we can merge :)

**Comment by lambday** [comment]
@karlnapf updated the PR with clang format changes. Thanks for the GPL clue. I didn't know we added it as a submodule. I was getting an error when configuring with GPL ON something like this
```
CMake Error at src/shogun/CMakeLists.txt:49 (MESSAGE):
   Shogun can only be built with GPL codes if the source files are in
   /home/lambday/Downloads/shogun/src/gpl.  Please download or disable with
   LICENSE_GPL_SHOGUN=OFF
```
Now with the submodule downloaded, it works perfectly.

**Comment by karlnapf** [comment]
Still style errors:
https://travis-ci.org/shogun-toolbox/shogun/jobs/337145184#L714

But windows passed, and all unit tests as well. Will merge

**Pull request merged**